packages/mobile/src/navigator/NavigationService.ts

```diff
@@ -43,7 +47,12 @@ export function replace(routeName: string, params?: object) {
     })
 }
 
-export function navigate(routeName: string, params?: object) {
+export function navigate<RouteName extends keyof StackParamList>(
+  ...args: undefined extends StackParamList[RouteName]
+    ? [RouteName] | [RouteName, StackParamList[RouteName]]
+    : [RouteName, StackParamList[RouteName]]
+) {
```

Comment:
This is the TypeScript magic which checks the right params are passed!

```tsx
// with the following config:
type StackParamList = {
  [Screens.MyRoute]: undefined
  [Screens.AnotherRoute]: { something: string }
  [Screens.MyWay]: { something: number } | undefined
}

// MyRoute doesn't take any param, it accepts both
navigate(Screens.MyRoute);
// and
navigate(Screens.MyRoute, undefined);

// AnotherRoute takes params, it only accepts the following and raises 
// a compile time error if one of the route param is missing or incorrect:
navigate(Screens.AnotherRoute, { something: "something" });
// This will also raise an error:
navigate(Screens.AnotherRoute)

// MyWay params can be undefined, it accepts both
navigate(Screens.MyWay, { something: 3 });
// and
navigate(Screens.MyWay)
// but errors with:
navigate(Screens.MyWay, { });
```